File: POTD/Binary Search/Cpp/1665. Minimum Initial Energy to Finish Tasks.cpp
// Using BINARY SEARCH ON ANSWERS
// TIME : O(NLOGN) + oO(LOG(109))
// SPACE: O(N)

struct Node{
    int diff;
    int actual;
    int min;
};
class Solution {
private:
bool check(int energy,vector<Node>&res){
    for (auto &node : res) {
        if (energy < node.min)
            return false;

        energy -= node.actual;
    }

    return energy>=0;
}
public:
    int minimumEffort(vector<vector<int>>& tasks) {
        vector<Node> res;
        int maxi=-1;
        for(auto &vec:tasks){
            res.push_back({vec[1]-vec[0],vec[0],vec[1]});
            maxi=max(maxi,vec[1]);
        }


        sort(begin(res),end(res),[&](Node &a,Node&b){
            if(a.diff!=b.diff){
                // largest diff first
                return a.diff>b.diff;
            }
            
            if(a.min!=b.min){
                // largest min enerygy first
                return a.min>b.min;
            }
            return a.actual < b.actual; 
        });


        // use binary search on ans

        int low=maxi;
        int high=INT_MAX;
        int reqEnergy=-1;
        while(low<=high){
            int mid=low+(high-low)/2;

            if(check(mid,res)){
                reqEnergy=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }

        return reqEnergy;
    }
};

File: POTD/Binary Search/Python/1665. Minimum Initial Energy to Finish Tasks.py
// Using BINARY SEARCH ON ANSWERS
// TIME : O(NLOGN) + oO(LOG(109))
// SPACE: O(N)


class Solution:
    def check(self,energy:int,res:List[Tuple[int, int, int]])->bool:

        for _, actual, min_energy in res:
            if energy<min_energy:
                return False
            
            energy-=actual
        
        return energy>=0
    def minimumEffort(self, tasks: List[List[int]]) -> int:
        res=[]
        maxi=-1
        for task in tasks:
            res.append((task[1]-task[0],task[0],task[1]))
            maxi=max(maxi,task[1])
        
        res.sort(
            key=lambda x : (-x[0], -x[2], x[1])
        )

        low=maxi
        high=10**9
        reqMinEnergy=-1

        while low<=high:
            mid=low + (high-low)//2

            if self.check(mid,res):
                reqMinEnergy=mid
                high=mid-1
            else:
                low=mid+1
        
        return reqMinEnergy
